python/task_manager/results_poller.py

```diff
@@ -9,103 +9,179 @@
+            # in order to implement a fire and forget scheme, which would be ideal.
+            # However, this wouldn't bring a lot of advantages. Since the only communication
+            # channel between the dispatcher and the backgroud manager is the self._fn
+            # variable, it means that the access to that variable must be made thread-safe
+            # so that it is not updated before the event is executed in the main thread.
+            # Doing so would require adding locks and complicate the code further.
+            # Using the builtin lock from the invoker here is much easier to understand.
+
+            # We could also add the events to invoke into a second queue meant to
+            # be consumed by the background task manager, but it would only add
+            # complexity to the design and it wouldn't provide any significant speed gain.
+            QtCore.QMetaObject.invokeMethod(self, "_do_invoke", QtCore.Qt.BlockingQueuedConnection)
+
+    @QtCore.Slot()
+    def _do_invoke(self):
+        """
+        Executes the event to dispatch.
+        """
         try:
-            # Get everything until the queue is empty.
-            while True:
-                # Get the next result from the queue
-                result_tuple = self._results.get_nowait()
-                try:
-                    # Result tuples with 3 values are coming from succesful tasks.
-                    if len(result_tuple) == 3:
-                        self.task_completed.emit(*result_tuple)
-                    else:
-                        # Result tuples with 4 values are coming from succesful tasks.
-                        self.task_failed.emit(*result_tuple)
-                except Exception:
-                    self._bundle.log_exception(
-                        "Exception thrown while reporting completed task."
-                    )
-                    # Do not re-raise, simply process the remaining events.
-        except Queue.Empty:
-            # Queue is empty, nothing to do!
-            pass
-
-    def queue_task_completed(self, worker_thread, task, result):
-        """
-        Called by background threads to notify that a task has completed.
-
-        :param worker_thread: Thread that completed the task.
-        :param task: Task that was completed.
-        :param result: Result produced by the thread.
-        """
-        self._results.put(
-            (worker_thread, task, result)
-        )
-
-    def queue_task_failed(self, worker_thread, task, msg, traceback_):
-        """
-        Called by background threads to notify that a task has failed.
-
-        :param worker_thread: Thread that completed the task.
-        :param task: Task that failed.
-        :param msg: Error message from the task,
-        :param traceback_: Call stack from the task.
-        """
-        self._results.put(
-            (worker_thread, task, msg, traceback_)
-        )
+            self._event.execute()
```

Comment:
Wondering if the generic event classes with their `execute()` method is making the code harder to follow? Could it just be queued up as a simple dictionary instead? Would that make it easier to follow? I think that it's the fact that all `execute()` does is to emit a signal from this class that makes me wonder if we need all the extra object wrapping?